### PR TITLE
ARROW-4065: [C++] arrowTargets.cmake is broken

### PR DESCRIPTION
**Pull request by sarutak**
When we build Arrow's cpp library using CMake, arrowTargets.cmake will be generated and installed but it's broken.

The following is a part of arrowTargets.cmake generated.

```
# Create imported target arrow_shared
add_library(arrow_shared SHARED IMPORTED)

set_target_properties(arrow_shared PROPERTIES
  INTERFACE_LINK_LIBRARIES "dl;pthreadshared"
)

# Create imported target arrow_static
add_library(arrow_static STATIC IMPORTED)

set_target_properties(arrow_static PROPERTIES
  INTERFACE_LINK_LIBRARIES "glog_static;zstd_static;zlib_shared;snappy_static;lz4_static;brotli_dec_static;brotli_enc_static;brotli_common_static;double-conversion_static;boost_system_shared;boost_filesystem_shared;boost_regex_shared;jemalloc_static;rt;pthreadshared"
)
```

There are no INTERFACE_INCLUDE_DIRECTORIES and linker doesn't recognize pthreadshared because the true name of pthread should be libpthread.so or libpthread.a.
*_static and *_shared are also wrong name.

After this fix, we can build apps which links to arrow using CMake with CMakeLists.txt like as follows.

```
cmake_minimum_required(VERSION ...)
project(...)
...
find_package(arrow)
add_executable(your_excellent_app ...)
target_link_libraries(your_excellent_app arrow_shared) # or arrow_static
...
```

`$ cmake -D CMAKE_PREFIX_PATH=/path/to/arrow /path/to/CMakeLists.txt`
`$ cmake --build .`

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -32,7 +32,7 @@ project(arrow VERSION "${ARROW_BASE_VERSION}")
 set(ARROW_VERSION_MAJOR "${arrow_VERSION_MAJOR}")
 set(ARROW_VERSION_MINOR "${arrow_VERSION_MINOR}")
 set(ARROW_VERSION_PATCH "${arrow_VERSION_PATCH}")
-if (ARROW_VERSION_MAJOR STREQUAL "" OR
+if(ARROW_VERSION_MAJOR STREQUAL "" OR
```
Comment (pitrou) [inline cpp/CMakeLists.txt]: I'm not fond of the style changes here. If we want to reformat the cmake files we should do so consistently in a separate PR, IMHO. Here it's making it more tedious to follow your changes.

(also it seems there are automatic tools for that, such as https://github.com/cheshirekow/cmake_format)

File: cpp/cmake_modules/BuildUtils.cmake [hunk 2]
```diff
@@ -24,62 +24,29 @@ function(ADD_THIRDPARTY_LIB LIB_NAME)
...
+  if(ARG_STATIC_LIB)
+    add_library(${LIB_NAME} STATIC IMPORTED)
+    set_target_properties(${LIB_NAME}
       PROPERTIES IMPORTED_LOCATION "${ARG_STATIC_LIB}")
     if(ARG_DEPS)
-      set_target_properties(${AUG_LIB_NAME}
+      set_target_properties(${LIB_NAME}
         PROPERTIES INTERFACE_LINK_LIBRARIES "${ARG_DEPS}")
     endif()
-    message("Added static library dependency ${AUG_LIB_NAME}: ${ARG_STATIC_LIB}")
```
Comment (pitrou) [inline cpp/cmake_modules/BuildUtils.cmake]: I don't understand why you removed all this.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -366,10 +360,10 @@ if (ARROW_BOOST_VENDORED)
     "${BOOST_LIB_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}boost_filesystem${CMAKE_STATIC_LIBRARY_SUFFIX}")
   set(BOOST_STATIC_REGEX_LIBRARY
     "${BOOST_LIB_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}boost_regex${CMAKE_STATIC_LIBRARY_SUFFIX}")
-  set(BOOST_SYSTEM_LIBRARY boost_system_static)
-  set(BOOST_FILESYSTEM_LIBRARY boost_filesystem_static)
-  set(BOOST_REGEX_LIBRARY boost_regex_static)
-  if (ARROW_BOOST_HEADER_ONLY)
+  set(BOOST_SYSTEM_LIBRARY boost_system)
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Can you explain why you're removing the "static" and "shared" prefixes everywhere? We're explicitly selecting one or the other variant of a library (for example, here we're linking statically with boost, while later below we're linking dynamically with it).

**Review by pitrou — DISMISSED** [review]
Right now the PR is difficult to evaluate due to the style changes.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 3 above)
Comment (sarutak) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: I understand that sub-projects may choose different linkage type (static or shared) for a library.
I'll submit another solution.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (sarutak) [reply cpp/CMakeLists.txt]: O.K, Let's discuss the style stuff in another PR.

**Comment by pitrou** [comment]
@sarutak Well, the problem is I don't understand what it's doing. You are introducing two new CMake variables without commenting on what they represent.

**Comment by sarutak** [comment]
@pitrou  <LIBNAME>Targets.cmake (or *-targets.cmake) is referred by CMake when applications which links to a library named <LIBNAME> is built by cmake command.

Arrow also provides arrowTargets.cmake so this means CMake user expect they can  build applications linking to Arrow like as follows.

```
$ cmake -DCMAKE_PREFIX_PATH=/path/to/arrow /path/to/target-application
```

But there are two problems.
First, there are no `INTERFACE_INCLUDE_DIRECTORIES` properties settings in `set_target_properties` command in arrowTargets.cmake.
Due to this issue, CMake can't find `include` directory for Arrow during target application compilation time.
arrowTargets.cmake is generated by CMake when we build Arrow. So to add `INTERFACE_INCLUDE_DIRECTORIES` properties, we need to tell where the include directory is by adding `INCLUDES DESTINATION` property to `install` command in `BuildUtils.cmake`.

The second problem is the library names following `INTERFACE_LINK_LIBRARIES` property in `set_target_properties` command are broken.
They are like `zlib_shared` or `pthreadshared`. They are  just target names, not library names.
Due to this issue, linking will fail during the target application is built. They should be like `z` and `pthread (or -lpthread)`.
`INTERFACE_LINK_LIBRARIES` property and values are also generated automatically so the simplest way to set library names to `INTERFACE_LINK_LIBRARIES` without the suffixes (like `_shared or _static`, set target library names to their library names in `target_link_libraries`.
But unlike most C++ libraries, Arrow needs to do `target_link_library` for both static and shared libraries in `ADD_THIRDPARTY_LIB` in `BuildUtils.cmake` (so** I understand my first solution is rejected).
Another solution is modifying `target_link_libraries` command for `libarrow` (not other libraries linked by Arrow).
By adding "$<BUILD_INTERFACE:${ARG_SHARED_LINK_LIBS}>", `"$<BUILD_INTERFACE:${ARG_STATIC_LINK_LIBS}>"` and `"$<INSTALL_INTERFACE:${INTERFACE_LIBS}>"` to the `target_link_libraries` command, we can refer the dependent library names as `*_static` or `_shared` while we can add suffix-less names to `INTERFACE_LINK_LIBRARIES` in arrowTargets.cmake to be generated.
The new variables added are for this.

**Comment by pitrou** [comment]
My main concern here is that maintaining the build system seems to become a specialist's thing. I'm afraid I have to defer reviewing this PR to someone more specialized than me.

**Review by xhochy — APPROVED** [review]
+1, I'm happy with this solution. I'm wondering whether we can add a test that checks that this is working.

When we have nightly conda packages again, we could also test this in https://github.com/wolfv/xtensor-arrow maybe.

**Comment by wesm** [comment]
I will take a look also since I've worked a lot on the CMake things

**Review by wesm — APPROVED** [review]
+1, thank you @sarutak -- we should think about how to make the interface link libraries easier to maintain. We have some other libraries in this project and they'll need some work also probably

**Comment by sarutak** [comment]
@wesm Yeah, I could contribute around that.
